docs/programmers_guide/witness_formal_spec.md

```diff
@@ -106,82 +269,126 @@ Helper functions MUST have at least one argument.
 Helper functions MAY have variadic parameters: `HELPER_EXAMPLE(arg1, arg2, list...)`.
```

Comment:
"Helper functions MUST be pure." Do we need to say that? Is it possible for them to be unpure?

Reply:
well, if you speak pseudocode -- no, but the real implementation is very easy to make unpure, so I think this should stay.

though, if a helper function logs something, it is unpure if we are being precise :-D